### PR TITLE
[CI] SearchWithRandomIOExceptionsIT#testRandomDirectoryIOExceptions  fix

### LINKED ISSUE #99174: [CI] SearchWithRandomIOExceptionsIT testRandomDirectoryIOExceptions failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/vugcgceftcmhk/tests/:server:internalClusterTest/org.elasticsearch.search.basic.SearchWithRandomIOExceptionsIT/testRandomDirectoryIOExceptions

**Reproduction line:**
```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.basic.SearchWithRandomIOExceptionsIT.testRandomDirectoryIOExceptions" -Dtests.seed=6B3F5FB7EAF0379E -Dtests.locale=vi-VN -Dtests.timezone=Africa/Monrovia -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.search.basic.SearchWithRandomIOExceptionsIT&tests.test=testRandomDirectoryIOExceptions

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=3707, name=elasticsearch[node_s1][generic][T#2], state=RUNNABLE, group=TGRP-SearchWithRandomIOExceptionsIT]


  Caused by: java.lang.AssertionError: java.lang.RuntimeException: MockDirectoryWrapper: cannot close: there are still 1 open files: {_2.cfs=1}

    at __randomizedtesting.SeedInfo.seed([6B3F5FB7EAF0379E]:0)
    at org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:71)
    at org.elasticsearch.index.store.Store.decRef(Store.java:406)
    at org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$9(StoreRecovery.java:488)
    at org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:236)
    at org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:182)
    at org.elasticsearch.index.CompositeIndexEventListener.callListeners(CompositeIndexEventListener.java:278)
    at org.elasticsearch.index.CompositeIndexEventListener.iterateBeforeIndexShardRecovery(CompositeIndexEventListener.java:287)
    at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:314)
    at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1681)
    at org.elasticsearch.index.shard.StoreRecovery.internalRecoverFromStore(StoreRecovery.java:410)
    at org.elasticsearch.index.shard.StoreRecovery.recoverFromStore(StoreRecovery.java:92)
    at org.elasticsearch.index.shard.IndexShard.recoverFromStore(IndexShard.java:2338)
    at org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:73)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
    at java.lang.Thread.run(Thread.java:1623)

    Caused by: java.lang.RuntimeException: MockDirectoryWrapper: cannot close: there are still 1 open files: {_2.cfs=1}

      at org.apache.lucene.tests.store.MockDirectoryWrapper.close(MockDirectoryWrapper.java:876)
      at org.apache.lucene.store.FilterDirectory.close(FilterDirectory.java:111)
      at org.apache.lucene.store.FilterDirectory.close(FilterDirectory.java:111)
      at org.elasticsearch.index.store.Store$StoreDirectory.innerClose(Store.java:768)
      at org.elasticsearch.index.store.Store.closeInternal(Store.java:436)
      at org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:117)
      at org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:69)
      at org.elasticsearch.index.store.Store.decRef(Store.java:406)
      at org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$9(StoreRecovery.java:488)
      at org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:236)
      at org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:182)
      at org.elasticsearch.index.CompositeIndexEventListener.callListeners(CompositeIndexEventListener.java:278)
      at org.elasticsearch.index.CompositeIndexEventListener.iterateBeforeIndexShardRecovery(CompositeIndexEventListener.java:287)
      at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:314)
      at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1681)
      at org.elasticsearch.index.shard.StoreRecovery.internalRecoverFromStore(StoreRecovery.java:410)
      at org.elasticsearch.index.shard.StoreRecovery.recoverFromStore(StoreRecovery.java:92)
      at org.elasticsearch.index.shard.IndexShard.recoverFromStore(IndexShard.java:2338)
      at org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:73)
      at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
      at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
      at java.lang.Thread.run(Thread.java:1623)

      Caused by: java.lang.RuntimeException: unclosed IndexInput: _2.cfs

        at org.apache.lucene.tests.store.MockDirectoryWrapper.addFileHandle(MockDirectoryWrapper.java:783)
        at org.apache.lucene.tests.store.MockDirectoryWrapper.openInput(MockDirectoryWrapper.java:835)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101)
        at org.apache.lucene.codecs.lucene90.Lucene90CompoundReader.<init>(Lucene90CompoundReader.java:78)
        at org.apache.lucene.codecs.lucene90.Lucene90CompoundFormat.getCompoundReader(Lucene90CompoundFormat.java:86)
        at org.apache.lucene.index.SegmentCoreReaders.<init>(SegmentCoreReaders.java:103)
        at org.apache.lucene.index.SegmentReader.<init>(SegmentReader.java:92)
        at org.apache.lucene.index.ReadersAndUpdates.getReader(ReadersAndUpdates.java:180)
        at org.apache.lucene.index.ReadersAndUpdates.getReadOnlyClone(ReadersAndUpdates.java:222)
        at org.apache.lucene.index.IndexWriter.lambda$getReader$0(IndexWriter.java:536)
        at org.apache.lucene.index.StandardDirectoryReader.open(StandardDirectoryReader.java:138)
        at org.apache.lucene.index.IndexWriter.getReader(IndexWriter.java:598)
        at org.apache.lucene.index.DirectoryReader.open(DirectoryReader.java:112)
        at org.apache.lucene.index.DirectoryReader.open(DirectoryReader.java:91)
        at org.elasticsearch.index.engine.InternalEngine.createReaderManager(InternalEngine.java:731)
        at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:283)
        at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:223)
        at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:14)
        at org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:2005)
        at org.elasticsearch.index.shard.IndexShard.openEngineAndRecoverFromTranslog(IndexShard.java:1969)
        at org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$9(StoreRecovery.java:479)
        at org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:236)
        at org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:182)
        at org.elasticsearch.index.CompositeIndexEventListener.callListeners(CompositeIndexEventListener.java:278)
        at org.elasticsearch.index.CompositeIndexEventListener.iterateBeforeIndexShardRecovery(CompositeIndexEventListener.java:287)
        at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:314)
        at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1681)
        at org.elasticsearch.index.shard.StoreRecovery.internalRecoverFromStore(StoreRecovery.java:410)
        at org.elasticsearch.index.shard.StoreRecovery.recoverFromStore(StoreRecovery.java:92)
        at org.elasticsearch.index.shard.IndexShard.recoverFromStore(IndexShard.java:2338)
        at org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:73)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.lang.Thread.run(Thread.java:1623)

```

### PR DESCRIPTION
**Pull request by pmpailis**
This PR aims to fix some test failures in `SearchWithRandomIOExceptionsIT#testRandomDirectoryIOExceptions` (https://github.com/elastic/elasticsearch/issues/99174). Looking at Gradle, it seems that the test started (re)failing on Aug 31st - and we kept having errors until it was muted. The exceptions raised were like the following (always with a `*cfs` file):
```
2> REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.basic.SearchWithRandomIOExceptionsIT.testRandomDirectoryIOExceptions" -Dtests.seed=664EB163B906E74D -Dtests.locale=en-GR -Dtests.timezone=Europe/Athens -Druntime.java=21
  2> java.lang.AssertionError: java.lang.RuntimeException: MockDirectoryWrapper: cannot close: there are still 1 open files: {_3.cfs=1}
        at __randomizedtesting.SeedInfo.seed([664EB163B906E74D]:0)
        at org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:71)
        at org.elasticsearch.index.store.Store.decRef(Store.java:407)
        at org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$9(StoreRecovery.java:488)
        at org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:236)
        at org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:182)
        at org.elasticsearch.index.CompositeIndexEventListener.callListeners(CompositeIndexEventListener.java:278)
        at org.elasticsearch.index.CompositeIndexEventListener.iterateBeforeIndexShardRecovery(CompositeIndexEventListener.java:287)
        at org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:314)
        at org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1683)
        at org.elasticsearch.index.shard.StoreRecovery.internalRecoverFromStore(StoreRecovery.java:410)
        at org.elasticsearch.index.shard.StoreRecovery.recoverFromStore(StoreRecovery.java:92)
        at org.elasticsearch.index.shard.IndexShard.recoverFromStore(IndexShard.java:2340)
        at org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:73)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        Caused by:
        java.lang.RuntimeException: MockDirectoryWrapper: cannot close: there are still 1 open files: {_3.cfs=1}
            at org.apache.lucene.tests.store.MockDirectoryWrapper.close(MockDirectoryWrapper.java:882)
            at org.apache.lucene.store.FilterDirectory.close(FilterDirectory.java:111)
            at org.apache.lucene.store.FilterDirectory.close(FilterDirectory.java:111)
            at org.elasticsearch.index.store.Store$StoreDirectory.innerClose(Store.java:769)
            at org.elasticsearch.index.store.Store.closeInternal(Store.java:437)
            at org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:117)
            at org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:69)
            ... 17 more
            Caused by:
            java.lang.RuntimeException: unclosed IndexInput: _3.cfs
                at org.apache.lucene.tests.store.MockDirectoryWrapper.addFileHandle(MockDirectoryWrapper.java:780)
                at org.apache.lucene.tests.store.MockDirectoryWrapper.openInput(MockDirectoryWrapper.java:832)
                at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101)
                at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101)
                at org.apache.lucene.codecs.lucene90.Lucene90CompoundReader.<init>(Lucene90CompoundReader.java:78)
                at org.apache.lucene.codecs.lucene90.Lucene90CompoundFormat.getCompoundReader(Lucene90CompoundFormat.java:86)
                at org.apache.lucene.index.SegmentCoreReaders.<init>(SegmentCoreReaders.java:103)
                at org.apache.lucene.index.SegmentReader.<init>(SegmentReader.java:92)
                at org.apache.lucene.index.ReadersAndUpdates.getReader(ReadersAndUpdates.java:180)
                at org.apache.lucene.index.ReadersAndUpdates.getReadOnlyClone(ReadersAndUpdates.java:222)
                at org.apache.lucene.index.IndexWriter.lambda$getReader$0(IndexWriter.java:536)
                at org.apache.lucene.index.StandardDirectoryReader.open(StandardDirectoryReader.java:138)
                at org.apache.lucene.index.IndexWriter.getReader(IndexWriter.java:598)
                at org.apache.lucene.index.DirectoryReader.open(DirectoryReader.java:112)
                at org.apache.lucene.index.DirectoryReader.open(DirectoryReader.java:91)
                at org.elasticsearch.index.engine.InternalEngine.createReaderManager(InternalEngine.java:766)
                at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:287)
                at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:226)
                at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:14)
                at org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:2007)
                at org.elasticsearch.index.shard.IndexShard.openEngineAndRecoverFromTranslog(IndexShard.java:1971)
                at org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$9(StoreRecovery.java:479)
                ... 15 more
``` 
The exception wasn't always reproducible, as while the random seeds were set correctly, the order at which the files were processed and we would determine whether to throw an IO exception wasn't guaranteed. 

After some investigation, this traced back to the changes introduced [in this PR](https://github.com/elastic/elasticsearch/pull/99069), as (probably accidentally) the order of the method calls changed, so now we were facing a scenario where the `store.readLastCommittedSegmentsInfo` might throw a random IO exception, and the `internalReaderManager` (that holds the `directoryReader` and its open handles) would be null. So, when reaching the `final` block, there wouldn't be an object to call `close` upon, leaving the open files from `DirectoryReader.open(indexWriter)` untouched. 

The suggested fix here is to simply restore the order of the operations, so that if an exception is raised from any downstream channel, the `directoryReader` would properly close all its references & open handles. 

@henningandersen  could you please also take a look and advise? Is this something that is limited to test code or could it also affect production (and should we backport)? 🙏 

Closes https://github.com/elastic/elasticsearch/issues/99174

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -766,8 +766,8 @@ private ExternalReaderManager createReaderManager(RefreshWarmerListener external
                     DirectoryReader.open(indexWriter),
                     shardId
                 );
-                lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:769]: Unfortunately the change here was quite deliberate in that we need to access the `lastCommittedSegmentInfos` when creating the reader manager in stateless.

Can we instead fix the exception handling here?

**Review by henningandersen — COMMENTED** [review]
I think we need to work on the exception handling instead, since the order is important.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -759,10 +759,11 @@ private static String loadHistoryUUID(Map<String, String> commitData) {
 
     private ExternalReaderManager createReaderManager(RefreshWarmerListener externalRefreshListener) throws EngineException {
         boolean success = false;
+        ElasticsearchDirectoryReader directoryReader = null;
```
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:762]: Removed the `final` here to have the reference & reassign later.

**Comment by pmpailis** [comment]
Updated error handling to close the initial reference to the directory if the `internalReaderManager` has not been initialized. @henningandersen  could you please have another look ? 🙇

**Review by henningandersen — APPROVED** [review]
LGTM

Thanks for fixing this.

**Pull request merged**